projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/asm/syscall/LLVMAMD64File.java

```diff
@@ -0,0 +1,337 @@
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE
+ * GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ * NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package com.oracle.truffle.llvm.nodes.asm.syscall;
+
+import java.io.File;
+import java.io.FileDescriptor;
+import java.io.FileInputStream;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.RandomAccessFile;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+
+import com.oracle.truffle.api.CompilerDirectives.TruffleBoundary;
+import com.oracle.truffle.llvm.nodes.asm.support.LLVMAMD64String;
+import com.oracle.truffle.llvm.runtime.LLVMAddress;
+import com.oracle.truffle.llvm.runtime.memory.LLVMMemory;
+
+import sun.misc.SharedSecrets;
+
+public class LLVMAMD64File {
+    public static final int O_ACCMODE = 00000003;
+    public static final int O_RDONLY = 00000000;
+    public static final int O_WRONLY = 00000001;
+    public static final int O_RDWR = 00000002;
+    public static final int O_CREAT = 00000100;
+    public static final int O_EXCL = 00000200;
+    public static final int O_NOCTTY = 00000400;
+    public static final int O_TRUNC = 00001000;
+    public static final int O_APPEND = 00002000;
+    public static final int O_NONBLOCK = 00004000;
+    public static final int O_DSYNC = 00010000;
+    public static final int FASYNC = 00020000;
+    public static final int O_DIRECT = 00040000;
+    public static final int O_LARGEFILE = 00100000;
+    public static final int O_DIRECTORY = 00200000;
+    public static final int O_NOFOLLOW = 00400000;
+    public static final int O_NOATIME = 01000000;
+    public static final int O_CLOEXEC = 02000000;
+    public static final int O_TMPFILE = 020000000;
+
+    public static final int SEEK_SET = 0;
+    public static final int SEEK_CUR = 1;
+    public static final int SEEK_END = 2;
+
+    private static boolean test(int flags, int flag) {
+        return (flags & flag) == flag;
+    }
+
+    private static FileDescriptor getFd(int fd) {
+        FileDescriptor f = new FileDescriptor();
+        SharedSecrets.getJavaIOFileDescriptorAccess().set(f, fd);
+        return f;
+    }
+
+    private static FileChannel getInputChannel(int fd) {
+        return new FileInputStream(getFd(fd)).getChannel();
+    }
+
+    private static FileChannel getOutputChannel(int fd) {
+        return new FileOutputStream(getFd(fd)).getChannel();
+    }
+
+    @TruffleBoundary
+    public static int open(String filename, int flags, @SuppressWarnings("unused") int mode) {
```

Comment:
I think it would be better if we put each method in a separate node. Then we could use profiling, for example, here for `rdwr`. Additionally, we could use `TruffleBoundary` on a more fine-grained basis to prevent, for example, that `LLVMAddress` has to be allocated on the heap. This applies not only to this class, but also to most of the other operation which could be implemented as nodes.